### PR TITLE
eeprom: at24: Add support for SMBus

### PR DESCRIPTION
**Pull request by finikorg**
-  Add support for SMBus for at24 driver, support trivial one page EEPROM found in Qemu Q35.
-  Add EEPROMs to Qemu x86_64 dts and include them to tests in `tests/drivers/eeprom/api/`

### CODE REVIEW

File: dts/bindings/mtd/atmel,at24-smbus.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2023 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: Atmel AT24 - compatible SMBus EEPROM
+
+compatible: "atmel,at24"
+
+include: ["atmel,at2x-base.yaml", smbus-device.yaml]
```
Comment (henrikbrixandersen) [inline dts/bindings/mtd/atmel,at24-smbus.yaml:9]: This is yet another annoyance with the way we ended up add SMbus support - now we need different bindings for exactly the same device on what is essentially the same type of bus, but with a different API.

I am still not sold on this at all, sorry.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: boards/x86/qemu_x86/qemu_x86_64.dts [hunk 2]
```diff
@@ -16,5 +16,93 @@
...
+			reg = <0x56>;
+			size = <0x100>;
+			pagesize = <0x100>;
+			address-width = <8>;
+			timeout = <5>;
+
+			status = "okay";
+		};
+
+		eeprom9: smbus_eeprom@57 {
```
[the remark this answers is not in the document]
Comment (finikorg) [reply boards/x86/qemu_x86/qemu_x86_64.dts:97]: I can, probably, move them to the test overlay. I just added all devices supported by Qemu.

File: dts/bindings/mtd/atmel,at24-smbus.yaml (hunk 1 above)
Comment (henrikbrixandersen) [reply dts/bindings/mtd/atmel,at24-smbus.yaml:9]: My concern is duplication efforts for I2C devices on a SMBus, which is basically just a restricted protocol on top of I2C. I am questioning if the way we decided to implement SMBus support in Zephyr, which leads to these duplication efforts, is the correct architecture.

Comment (finikorg) [reply dts/bindings/mtd/atmel,at24-smbus.yaml:9]: What is exactly "duplicated effort"? ;) Did I add 100% more code to the driver?

Comment (henrikbrixandersen) [reply dts/bindings/mtd/atmel,at24-smbus.yaml:9]: Duplicated, as in you needed to modify the binding and driver at all. This is the exact same device on the same type of bus, but with restrictions. This shows that _each_ driver, which a user need to support on the SMBus, requires driver changes.

From a quick grep in the source tree, we have 160 I2C device drivers. In order to support these devices on an SMBus, each of these bindings and drivers would need to be modified.

Comment (finikorg) [reply dts/bindings/mtd/atmel,at24-smbus.yaml:9]: What exact `each driver` are we talking about here? So far there is no use case AFAIK. Even current at24 driver is for qemu for testing purposes only.
Real EEPROM driver on SMBus would need to switch pages using specific smbus commands like it is done in Linux:
https://github.com/torvalds/linux/blob/44c026a73be8038f03dbdeef028b642880cf1511/drivers/misc/eeprom/ee1004.c#L83

Comment (henrikbrixandersen) [reply dts/bindings/mtd/atmel,at24-smbus.yaml:9]: Then I am even more puzzled about this PR. Why not add a "real" SMBus EEPROM driver, then?

Comment (finikorg) [reply dts/bindings/mtd/atmel,at24-smbus.yaml:9]: OK, I close PR since I used it only for testing with Qemu

**Comment by finikorg** [comment]
Closing request, as the driver is used only for testing with Qemu, no real device.

**Pull request closed without merging**